txpipe/covariance.py

```diff
@@ -61,206 +77,419 @@ def read_cosmology(self):
+                Bmode_F=-1 
+            cov[1324] += np.outer(cl[13]*0 + SN[13], cl[24]*0 + SN[24]) * coupling_mat[1324] * Bmode_F
+            cov[1423] += np.outer(cl[14]*0 + SN[14], cl[23]*0 + SN[23]) * coupling_mat[1423] * Bmode_F
+
+        cov['final']=cov[1423]+cov[1324]
+
+        if self.do_xi:
+            s1_s2_1 = self.get_spins(tracer_comb1)
+            s1_s2_2 = self.get_spins(tracer_comb2)
+
+            # For the shear-shear we have two sets of spins, plus and minus,
+            # which are returned as a dict, so we need to pull out the one we need
+            # Otherwise it's just specified as a tuple, e.g. (2,0)
+            if isinstance(s1_s2_1, dict):
+                s1_s2_1 = s1_s2_1[xi_plus_minus1]
+            if isinstance(s1_s2_2, dict):
+                s1_s2_2 = s1_s2_2[xi_plus_minus2]
+
+            # Use these terms to project the covariance from C_ell to xi(theta)
+            th, cov['final']=WT.projected_covariance2(
+                l_cl=ell, s1_s2=s1_s2_1, s1_s2_cross=s1_s2_2, cl_cov=cov['final'])
+
+        # Normalize
+        cov['final'] /= norm
+
+        # Put the covariance into bins. 
+        # This is optional in the case of a C_ell covariance (only if bins in ell are
+        # supplied, otherwise the matrix is for each ell value individually).  It is
+        # required for real-space covariances since these are always binned.
+        if self.do_xi:
+            thb, cov['final_b'] = bin_cov(r=th/d2r, r_bins=ell_bins, cov=cov['final'])
+        else:
+            if ell_bins is not None:
+                lb, cov['final_b'] = bin_cov(r=ell, r_bins=ell_bins, cov=cov['final'])
+        return cov
+    
+    def get_angular_bins(self, two_point_data):
+        # Assume that the ell binning is the same for each of the bins.
+        # This is true in the current pipeline.
+        X = two_point_data.get_data_points('galaxy_shear_cl_ee',i=0,j=0)
+        # Further assume that the ell ranges are contiguous, so that
+        # the max value of one window is the min value of the next.
+        # So we just need the lower edges of each bin and then the
+        # final maximum value of the last bin
+        ell_edges = [x['window'].min for x in X]
+        ell_edges.append(X[-1]['window'].max)
+
+        return np.array(ell_edges)
+
+    #compute all the covariances and then combine them into one single giant matrix
+    def compute_covariance(self, cosmo, meta, two_point_data={}):
+        from tjpcov import bin_cov, wigner_transform
+        import threadpoolctl
+
+        #FIXME: Only input needed should be two_point_data, which is the sacc data file.
+        # Other parameters should be included within sacc and read from there.
+        ccl_tracers,tracer_Noise = self.get_tracer_info(cosmo, meta, two_point_data=two_point_data)
+        # we will loop over all these
+        tracer_combs = two_point_data.get_tracer_combinations() 
+        N2pt = len(tracer_combs)
+        
+        N2pt0 = -1
```

Comment:
I don't understand what this bit is doing, why the counter starts at -1, or why we are multiplying things by 1.  Can you add some explicatory comments?

Reply:
ok this bit is a bit of a hack, all it's doing is counting the number of 2pt functions and making sure xip xim are both counted... I added some comments but if you have a better way to do this please change